Split `test_testing.py::TestAsserts` for multiple devices

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56365 Split `test_testing.py::TestAsserts` for multiple devices**

Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

Differential Revision: [D28247529](https://our.internmc.facebook.com/intern/diff/D28247529)

pmeier committed:
Split `test_testing.py::TestAsserts` for multiple devices

Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

walterddr commented on test/test_testing.py:
```diff
@@ -1055,68 +1016,81 @@ def test_mapping_mismatching_values_msg(self, device):
+            fn(check_device=False)
+
+
+instantiate_device_type_tests(TestAssertsMultiDevice, globals())
```
nit: you can also do 
```suggestion
instantiate_device_type_tests(TestAssertsMultiDevice, globals(), only_for='cuda')
```
so you dont need to explicitly set `@onlyCUDA` decorator for each tests above

walterddr reviewed (approved):
looks good. thanks for the prompt follow up @pmeier

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

mruberry commented on test/test_testing.py:
```diff
@@ -745,7 +745,7 @@ def test_trivial_passing_test(self, device):
 
 
-class TestAsserts(TestCase):
+class _TestAssertsMixin:
```
I don't think we need a mixin for this?

mruberry replied:
I would just make these functions. They don't use self, right?

mruberry commented on test/test_testing.py:
```diff
@@ -813,187 +813,152 @@ def assertRaisesRegexs(self, expected_exception: Type[Exception], *expected_rege
-        actual = torch.ones(1, dtype=torch.float32, device=device)
-        expected = torch.ones(1, dtype=torch.complex64, device=device)
+
+class TestAsserts(TestCase, _TestAssertsMixin):
```
Keep this part of the device generic framework but instantiate its tests only for the CPU

mruberry commented on test/test_testing.py:
```diff
@@ -1055,68 +1016,79 @@ def test_mapping_mismatching_values_msg(self, device):
+            with self.assertRaisesRegex(AssertionError, "device"):
+                fn()
+
+    def test_mismatching_device_no_check(self, device):
```
Add a test with multiple cuda devices (cuda:0 and cuda:1, for example), too. We have a decorator for these tests:

https://github.com/pytorch/pytorch/blob/df8bb5a42b3c2b1c57bbec9a1c60db6aeafe0f20/torch/testing/_internal/common_device_type.py#L732

mruberry commented:
This refactor PR seems unnecessary but I made some comments if we really want to pursue it as part of this stack

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

[ghstack-poisoned]

mruberry commented on test/test_testing.py:
```diff
@@ -744,412 +744,392 @@ def test_trivial_passing_test(self, device):
+    return [
+        (actual, expected),
+        ((actual,), (expected,)),
+        ([actual], [expected]),
```
might as well do list x tuple, too

mruberry commented on test/test_testing.py:
```diff
@@ -744,412 +744,392 @@ def test_trivial_passing_test(self, device):
 
 
-instantiate_device_type_tests(TestAsserts, globals())
+def device_combinations():
```
This seems like a bit of overkill; just use the deviceCountAtLeast decorator on the test cases:

https://github.com/pytorch/pytorch/blob/866b19e95d1b28fee07161e92692d2cccc788dcd/torch/testing/_internal/common_device_type.py#L738

https://github.com/pytorch/pytorch/blob/866b19e95d1b28fee07161e92692d2cccc788dcd/torch/testing/_internal/common_device_type.py#L34

That can used to create the permutations, if you like

pmeier replied:
See https://github.com/pytorch/pytorch/pull/56365#discussion_r627138093.

mruberry reviewed (approved):
Overall looks really good, @pmeier; just a couple suggestions

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28247529).

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

Differential Revision: [D28247529](https://our.internmc.facebook.com/intern/diff/D28247529)

[ghstack-poisoned]

pmeier commented on test/test_testing.py:
```diff
@@ -744,413 +745,396 @@ def test_trivial_passing_test(self, device):
+
+
+class TestAssertsMultiDevice(TestCase):
+    @deviceCountAtLeast(1)
```
@mruberry I dislike this for two reasons:

1. We now need to slap the same decorator on every test in this test case.
2. Without knowing that `@deviceCountAtLeast(1)` returns a list of available devices rather than a single device this can be confusing: 
    > Why do I need this decorator if the test case is instatiated with `only_for="cuda"`? Doesn't that already check that CUDA is available and thus I have at least one device?"

mruberry replied:
It lets you get rid of the `device_permutations` helper because you can inline it as just

```
itertools.permutations(("cpu", *cuda_devices), 2)
```

mruberry replied:
Looking forward, too, it's important today that we understand if a test would use a hardware accelerator or not, tomorrow it might be important we understand if it could use multiple hardware accelerators, and this is decorator is the best way we have of understanding that programmatically

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

Differential Revision: [D28247529](https://our.internmc.facebook.com/intern/diff/D28247529)

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

Differential Revision: [D28247529](https://our.internmc.facebook.com/intern/diff/D28247529)

[ghstack-poisoned]

pmeier committed:
Update on "Split `test_testing.py::TestAsserts` for multiple devices"


Follow-up to https://github.com/pytorch/pytorch/pull/54784#discussion_r614156172. Instead of having one large testcase where most methods are decorated with `@onlyCPU`, this factors out all tests that actually need another device into a separate test case.

Differential Revision: [D28247529](https://our.internmc.facebook.com/intern/diff/D28247529)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28247529).
